### PR TITLE
Created ArangoDB connector

### PR DESCRIPTION
**Pull request by apfohl**
- Created SessionService
- Created ArangoService
- Created ArangoInterceptor
- Created models (Result, Jwt)
- Added basic UI elements to test the connector

In order to try this out you need to setup an ArangoDB on the default port. Have a look here: https://hub.docker.com/_/arangodb

### CODE REVIEW

**Comment by apfohl** [comment]
As further investigation reveals, it is not as easy as it might look to use the user management of the ArangoDB. Every user would have to have administrator privileges in order retrieve e.g. the list of all users. This is also true for CouchDB.

So, you can have a look at what I've done so far, but keep in mind that this does not make any sense.

We probably need a small backend layer providing a good abstraction of the database and handles the authentication.

**Comment by Kharonus** [comment]
Should we even merge this then back to develop or keep it in this feature branch.

Point for merging is we can possibly reuse some code for a connection to a backend. Point against could be that we are introducing the possibility of dead code here.

File: src/app/services/session.service.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public token(): string {
+        return this.webToken;
+    }
+
+    public register(username: string, password: string): void {
+        console.log('Register!');
+        this.httpClient.post<Jwt>(`${this.baseUrl}/_open/auth`, { username, password })
+            .subscribe((value) => {
+                this.webToken = value.jwt;
+                console.log(this.webToken);
```
Comment (Kharonus) [inline src/app/services/session.service.ts]: remove the console log

or do you see any reason for keeping it?

File: src/app/services/arango.service.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import { Observable } from 'rxjs';
+import { HttpClient } from '@angular/common/http';
+import { map } from 'rxjs/operators';
+
+import { Result } from '../models/result.model';
+
+@Injectable()
+export class ArangoService {
+
+    private readonly baseUrl = 'http://localhost:8529/_api';
```
Comment (Kharonus) [inline src/app/services/arango.service.ts]: Where does this naming scheme comes from for the url path? with the underscore? seems weird, is this arango specific?

File: src/app/models/jwt.model.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export class Jwt {
+
+    public jwt: string;
```
Comment (Kharonus) [inline src/app/models/jwt.model.ts]: is this web token class something, that is extend at some point? What is the purpose of having a class containing only a string property?

Also: I'd like to suggest a more meaningful name here, maybe even: JsonWebToken. At least in my understanding jwt is not as common as shortcut like id, param, etc.

File: src/app/services/session.service.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+@Injectable()
+export class SessionService {
+
+    private baseUrl = 'http://localhost:8529';
+
+    private webToken: string = null;
+
+    public constructor(private httpClient: HttpClient) {}
+
+    public token(): string {
```
Comment (Kharonus) [inline src/app/services/session.service.ts]: could this be a getter instead of a method?

File: src/app/app.component.ts [hunk 5]
```diff
@@ -1,10 +1,33 @@
...
+    public databases: Array<string> = [];
+
+    public constructor(private arangoService: ArangoService, private sessionService: SessionService) {}
+
+    public onSubmit(): void {
+        this.sessionService.register(this.username, this.password);
+    }
+
+    public sendRequest(): void {
+        console.log('Send request');
```
Comment (Kharonus) [inline src/app/app.component.ts]: remove console logs

**Comment by crs-** [comment]
> 
> 
> As further investigation reveals, it is not as easy as it might look to use the user management of the ArangoDB. Every user would have to have administrator privileges in order retrieve e.g. the list of all users. This is also true for CouchDB.

What about saving the useres in the DB itself instead of calling for all registered users?

**Comment by Kharonus** [comment]
> What about saving the useres in the DB itself instead of calling for all registered users?

Is this something that is common for this kind of databases? (Sorry, DB-New-Guy question)

**Comment by apfohl** [comment]
Then you need to ship admin credentials with the web application to access the database to receive the users. This wouldn't be better. An abstraction is necessary to keep the database safe. Further, the database does not need to be world visible, but just the API that is really needed.

**Comment by crs-** [comment]
> 
> 
> > What about saving the useres in the DB itself instead of calling for all registered users?
> 
> Is this something that is common for this kind of databases? (Sorry, DB-New-Guy question)

I'm not shure, but I remember that user management was only implemented superficial for ArangoDB. But AFAIK calling all registered users also needs admin rights in most other database systems. Because it stores the users in system databases.

**Comment by apfohl** [comment]
@crs- Exactly.

**Comment by crs-** [comment]
> 
> 
> Then you need to ship admin credentials with the web application to access the database to receive the users. This wouldn't be better. An abstraction is necessary to keep the database safe. Further, the database does not need to be world visible, but just the API that is really needed.

Did you play with the ArangoDB Foxx framework?

File: src/app/models/jwt.model.ts (hunk 3 above)
Comment (apfohl) [reply src/app/models/jwt.model.ts]: I just introduced this model to have the HTTP client automagically use it's types for `pipe` and other operators. Please have a look there.

File: src/app/services/arango.service.ts (hunk 2 above)
Comment (apfohl) [reply src/app/services/arango.service.ts]: It's ArangoDB HTTP API

File: src/app/services/session.service.ts (hunk 4 above)
Comment (apfohl) [reply src/app/services/session.service.ts]: In this case sure. I once had some more code in there.

**Comment by apfohl** [comment]
I did not, but had a quick look. We would have to write all the backend logic in plain JS. Not my favourite. But might be worth another look.

**Comment by apfohl** [comment]
@Kharonus I would not want to have this merged, but I could branch from this clear out the ArangoDB stuff and create another PR with the cleaned up foundation.

File: src/app/models/jwt.model.ts (hunk 3 above)
Comment (Kharonus) [reply src/app/models/jwt.model.ts]: Ah, I see. Then still I'd to get rid of the abbr. ;)

**Comment by apfohl** [comment]
@crs- No, you can't for example access the user management of ArangoDB if the user hasn't administrative privileges. So, you will still not be able to retrieve the list of users or link any expenses to   a user account. I guess we are trying to misuse the database for something that it's not meant to be used for.

**Comment by apfohl** [comment]
Ok, I this: http://devblog.ztp.pt/using-typescript-for-programming-arangodb-foxx-services/.

I will try to create a small FOXX service allowing some feature we might want to use.

**Comment by apfohl** [comment]
I fixed some suggestions from the reviews. Just to have this clean. But this will not get merged anyway.

**Comment by apfohl** [comment]
I’ve read along and found this one: 
https://docs.arangodb.com/3.4/Manual/Foxx/Guides/Browser.html.

> Only use this for internal services intended to help users who already have full access to the database. Don't ever expose your database to the public Internet.

I guess it’ll be better to write a backend. But we can still try to extract some logic into an ArangoDB micro service. E.g. the calculation part.

Any objections? If not, I will close this and we will talk about a backend.

**Comment by Kharonus** [comment]
No objections, m'lord!

**Pull request closed without merging**